### PR TITLE
User stats page

### LINKED ISSUE #33: Statistic for employee
Each employee should have a statistics page on which some of the following should be shown:

- Extra time earned/lost in total
- Vacation days used this year
- Upcoming holidays
- Filter down by year/month for all stats

### PR DESCRIPTION
**Pull request by Baxxx**
Closes #33 and #19. 
This page prints out statistics for employee (his total/monthly bonus minutes, used vacation/sick days, working on weekends/holidays, etc.
Review pull requests #37 and #31, as this branch uses/updates some code from those branches.

### CODE REVIEW

**Comment by Baxxx** [comment]
Function `stats` in user.rb needs to be optimized. Working on it.

**Comment by sls-lnluksa** [comment]
@Baxxx 
- [ ] route /users/:id/statistic redirect to other page
- [ ] Link that should take user to statistic also don't work

File: app/controllers/companies_controller.rb [hunk 1]
```diff
@@ -65,6 +65,27 @@ def all_employees
...
+  end
+
+  def off_days_unapproved
+    @company = Company.find(params[:id])
+    authorize @company
+
+    @offdays = OffDay.where(company_id: @company.id, approved: false).order(from: :desc)
+  end
+
+  def all_employees
```
Comment (sls-lnluksa) [inline app/controllers/companies_controller.rb]: This is duplicate. Delete one, and make sure that we have correct code.

File: app/controllers/off_days_controller.rb [hunk 2]
```diff
@@ -71,13 +71,27 @@ def destroy
...
+    else
+      redirect_to URI(request.referer).path, notice: "Off day unsuccessfully approved"
+    end
+  end
+
+  def disapprove
+    off_day = OffDay.find(params[:off_day_id])
+    authorize off_day
+    off_day.approved = nil
+    # off_day.update_attribute(approved: nil)
```
Comment (sls-lnluksa) [inline app/controllers/off_days_controller.rb]: Remove

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -177,6 +177,50 @@ def review_invitation
...
 
+  def statistics
+    @user = User.find(params[:id])
+    authorize @user
+
+
+    @upcoming_holidays = @user.current_company.upcoming_holidays
+    @upcoming_offdays = @user.upcoming_off_days
+
+    @months = [I18n.t("month.january"), I18n.t("month.february"), I18n.t("month.march"), I18n.t("month.april"), I18n.t("month.may"), I18n.t("month.june"), I18n.t("month.july"), I18n.t("month.august"), I18n.t("month.september"), I18n.t("month.october"), I18n.t("month.november"), I18n.t("month.december")]
```
Comment (sls-lnluksa) [inline app/controllers/users_controller.rb]: Try to get all month_names from .yml files. Not one by one but all at once.

File: app/models/time_table.rb [hunk 4]
```diff
@@ -1,38 +1,93 @@
...
-      if records.any?
-        total_minutes = records.first.total_minutes
-      else
-        total_minutes = 0
+      total_minutes = 0
+      if records.where(approved: true).any?
+        total_minutes += records.where(approved: true, multirecord: false).first.total_minutes
+      end
+
+      today_offdays = OffDay.where("user_id = ? AND company_id = ? AND approved = TRUE AND off_days.from <= ? and off_days.to >= ? ",
```
Comment (sls-lnluksa) [inline app/models/time_table.rb]: Move this out of for loop, and adjust logic so we don't have so many queries.

File: app/models/time_table.rb [hunk 5]
```diff
@@ -1,38 +1,93 @@
...
+                             user.id, company.id, date.to_date, date.to_date)
+
+      offday = false
+      offday = today_offdays.first.absence_type if today_offdays.any? and date.strftime("%A")[0]!='S'
+      if offday # FOR TRANSLATION
+        offday.downcase!
+        offday.tr!(" ","_")
+      end
+
+      today_holidays = company.all_holidays.where("date(date) = ? ", date.to_date)
```
Comment (sls-lnluksa) [inline app/models/time_table.rb]: Move out of for loop.

File: app/models/time_table.rb [hunk 6]
```diff
@@ -1,38 +1,93 @@
...
+
+      holiday = false
+      holiday = today_holidays.first.name if today_holidays.any?
+
+      if records.where(approved:true).any? and (today_offdays.any? or today_holidays.any?)
```
Comment (sls-lnluksa) [inline app/models/time_table.rb]: Check for weekends

File: app/models/user.rb [hunk 7]
```diff
@@ -32,6 +33,105 @@ def active_companies
...
+
+
+    stats[:normal_minutes] = all_records.where("date(start) IN (?)",
+                                        (days_when_user_worked-
+                                        days_when_it_was_holiday-
+                                        days_when_user_was_on_offday)&
+                                        monday_to_friday
+                                        ).sum(:total_minutes)
+
+    stats[:normal_minutes] -= stats[:not_on_work_days]*480
```
Comment (sls-lnluksa) [inline app/models/user.rb]: Add # TODO
This does not work correct. We may calculate minutes for days before user start to work.

Comment (Baxxx) [inline app/models/user.rb]: Fixed, by changing date range in all calculations.

**Pull request merged**